source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -50,11 +54,133 @@ void Filter::onRead() {
+    // Error check the source and destination fields. Most errors are caught by the address
+    // parsing above, but a malformed IPv6 address may combine with a malformed port and parse as
+    // an IPv6 address when parsing for an IPv4 address. Remote address refers to the source
+    // address.
+    const auto remote_version = proxyProtocolHeader_.value().remote_address_->ip()->version();
+    const auto local_version = proxyProtocolHeader_.value().local_address_->ip()->version();
+    if (remote_version != proxyProtocolHeader_.value().protocol_version_ ||
+        local_version != proxyProtocolHeader_.value().protocol_version_) {
+      throw EnvoyException("failed to read proxy protocol");
+    }
+    // Check that both addresses are valid unicast addresses, as required for TCP
+    if (!proxyProtocolHeader_.value().remote_address_->ip()->isUnicastAddress() ||
+        !proxyProtocolHeader_.value().local_address_->ip()->isUnicastAddress()) {
+      throw EnvoyException("failed to read proxy protocol");
+    }
+
+    // Only set the local address if it really changed, and mark it as address being restored.
+    if (*proxyProtocolHeader_.value().local_address_ != *socket.localAddress()) {
+      socket.setLocalAddress(proxyProtocolHeader_.value().local_address_, true);
+    }
+    socket.setRemoteAddress(proxyProtocolHeader_.value().remote_address_);
+  }
+
+  // Release the file event so that we do not interfere with the connection read events.
+  file_event_.reset();
+  cb_->continueFilterChain(true);
+}
+
+size_t Filter::lenV2Address(char* buf) {
+  const uint8_t proto_family = buf[PROXY_PROTO_V2_SIGNATURE_LEN + 1];
+  size_t len;
+
+  switch ((proto_family & 0xf0) >> 4) {
+  case PROXY_PROTO_V2_AF_INET:
+    len = PROXY_PROTO_V2_ADDR_LEN_INET;
+    break;
+  case PROXY_PROTO_V2_AF_INET6:
+    len = PROXY_PROTO_V2_ADDR_LEN_INET6;
+    break;
+  default:
+    throw EnvoyException("Unsupported V2 proxy protocol address family");
+  }
+  return len;
+}
+
+void Filter::parseV2Header(char* buf) {
+  // Skip the first 12-bytes
+  // Next is ver/cmd
+  const int ver_cmd = buf[PROXY_PROTO_V2_SIGNATURE_LEN];
+  uint8_t u = buf[PROXY_PROTO_V2_HEADER_LEN - 2];
+  uint8_t l = buf[PROXY_PROTO_V2_HEADER_LEN - 1];
+  size_t hdr_addr_len = (u << 8) + l;
+
+  // Only do connections on behalf of another user, not
+  // internally-driven health-checks. If its not on behalf
+  // of someone, or its not AF_INET{6} / STREAM/DGRAM, ignore
+  // and use the real-remote info
+  if ((ver_cmd & 0xf) == PROXY_PROTO_V2_ONBEHALF_OF) {
+    uint8_t proto_family = buf[PROXY_PROTO_V2_SIGNATURE_LEN + 1];
+    if ((((proto_family & 0xf0) >> 4) == PROXY_PROTO_V2_AF_INET ||
+         ((proto_family & 0xf0) >> 4) == PROXY_PROTO_V2_AF_INET6) &&
+        (((proto_family & 0x0f) == PROXY_PROTO_V2_TRANSPORT_STREAM) ||
+         ((proto_family & 0x0f) == PROXY_PROTO_V2_TRANSPORT_DGRAM))) {
+      if (((proto_family & 0xf0) >> 4) == PROXY_PROTO_V2_AF_INET) {
```

Comment:
This looks like it will allow through invalid families and protocols as long as it is using the PROXY command.

I think it could be simplified by doing this:

If it's a LOCAL command, return early since it ignores the address information.
If it's a PROXY command (ONBEHALF_OF), check for these three cases of protocol values:

`\x11` or `\x12` are IPv4.
`\x21` or `\x22` are IPv6.
`\x00`, `\x31` or `\x32` are UNSPEC (or UNIX which we're treating like unspec) and should use the real connection info.

Any other command or protocol value must be rejected.

Reply:
ok i made half this change. Its not a bitfield so i can't really check against 4 values in one shot w/o making a macro, which doesn't make it simpler IMHO.